docs/reference-manual/native-image/DebugInfo.md

```diff
@@ -394,6 +395,46 @@ end
 0x8779c8:	"[Ljava.lang.String;"
 ```
 
+The native image heap contains a unique hub object (i.e. instance of
+`java.lang.Class`) for every Java type that is included in the
+image. It is possible to refer to these class constants using the
+standard Java class literal syntax:
+
+```
+(gdb) print 'Hello.class'
+$6 = {
+  <java.lang.Object> = {
+    <_objhdr> = {
+      hub = 0xaabd00,
+      idHash = 1589947226
+    }, <No data fields>}, 
+  members of java.lang.Class:
+  typeCheckStart = 13,
+  name = 0xbd57f0,
+  ...
+```
+
+Unfortunately it is necessary to quote the class constant literal to
+avoid gdb interpreting the embedded `.` character as a field access.
+
+Note that the type of a class constant literal is `java.lang.Class`
+rather than `java.lang.Class *`.
+
+Class constants exist for Java instance classes, interfaces, array
+classes and arrays, including primitive arrays:
+
+```
+(gdb)  print 'java.util.List.class'.name
```

Comment:
@adinn, an alternative approach would be to add an artificial static `class` field to each type so that the hub of a given type can be accessed via
```
p *'java.util.List'::class
```
wouldn't that fit better to the existing code C++ based code model that we have?

Reply:
I believe the field would better be called `.class` so that it's not a valid Java bytecode field name, if DWARF permits that and it doesn't make problems with GDB, that is:
`'java.util.List'::.class`

Reply:
> @adinn, an alternative approach would be to add an artificial static class-field to each type so that the hub of a given type can be accessed via
> ```
> p *'java.util.List'::class
> ```
> wouldn't that fit better to the existing C++ based code model that we have?

@olpaw That is indeed a nicer fit. The only problem with it is that it doesn't work ;-).

The only real option we have to make this data value visible via debug info is to encode it as a variable/constant, whether at top level or as a (static) member of some class. Any corresponding info definition must identify where the variable is located in memory. So, a static field info definition for a pseudo-field as described above would need specify as location the address of a memory slot whose content is an `oop`, a reference to the `Class` instance. We can make that reference a direct or indirect( (encoded) address by using a suitable location expression. However, we cannot get rid of the need for there to be an indirection.

Unfortunately, the only thing we have available to encode as the location is the address of the `Class` instance. *That* is the reason why I have had to declare the type of constant `java.util.List.class` as `java.lang.Class` not `java.lang.Class *`. It's also why when using isolates the type remains `java.lang.Class` not `_z_.java.lang.Class` or `_z_.java.lang.Class *`

That said, it is still possible to embed this constant as a field of the class. The required change is simply to generate the declaration and definition with the scope of the class and adjust the name and linkage name. The inescapable constraint we are under is that the field has to be an object value not an object reference.

So, we can support either:
```
(gdb) p 'java.util.List'::class
```
or
```
(gdb) p 'java.util.List'::.class
```

n.b. the reason I chose to make the constant a top level value rather than a value embedded in the class was:

1. the syntax and use of name `java.util.List.class` is very much like that of its Java original i.e. it functions as a global name for the class (modulo the associated network of class loader spaces within which it is visible)
2. in Java the class object is not actually a field of the class so presenting it as such appeared to me to risk engendering unnecessary confusion

I'm very happy to change this if you still want me to after taking the above explanation into account.

@peter-hofer yYour worry about `Foo::class` clashing with a legitimate field name is unnecessary. `class` is a reserved keyword so it cannot be used as a field name.

Reply:
@adinn `class` is a keyword in the Java _language_, but I would be surprised if it's explicitly forbidden in Java bytecode and can therefore not be used by other languages and especially bytecode generating tools.

Reply:
@peter-hofer Oh yes, of course `class` could well be inserted into synthetic or transformed bytecode as a class pool constant. I'd discounted that as something to worry about but now that you mention it I suppose there is quite probably someone somewhere crazy enough to have decided to use it as as field name.

Mind you, the same concern would apply for `.class`, perhaps yet more so in that it is not a keyword and therefore more attractive to less crazy people ;-).

I actually think this is all the more reason to keep encoding the class constants as a top level name. All other Java names are going to exist within the scope of a class (which effectively means a namespace). I think it is relatively unlikely that something like `org.my.Foo.class` is going to turn up as a name for a variable in non-Java code. Of course, that is asking to be proven wrong ...

Reply:
@adinn `.class` is not a valid identifier in bytecode because `.` is not allowed in identifiers (JVMS 4.2.2), hopefully that rule is enforced too. `org.my.Foo.class` could again refer to a qualified class or interface named `class` in package `org.my.Foo` -- hopefully it never does, but it's not prohibited.

Reply:
@peter-hofer is right, it's trivial to end up with names that are reserved keywords in Java with bytecode producers other than `javac`. E.g.
```scala
object Hello {
    val `class`: String = "WHY NOT?"
    def main(args: Array[String]) = {
        println(`class`)
    }
}
```
Unfortunately I'm not sure about `.class` either, because the following also works just fine:
```scala
object Hello {
    val `.class`: String = "WHY NOT?"
    def main(args: Array[String]) = {
        println(`.class`)
    }
}
```
But it could be that here Scala applies some magic to make this possible.

Reply:
@olpaw, Scala seems to do just that because the field is actually named `$u002Eclass` when using reflection.

Reply:
```
object Hello {
  val `class`: String = "OH NO";
  val `.class`: String = "WHY NOT?";

  def main(args: Array[String]) = {
    for (f <- getClass.getDeclaredFields) {
      println(f.getName())
    }
  }
}
```
prints
```
class
$u002Eclass
MODULE$
```